application/indexer/value.go

```diff
@@ -93,53 +99,74 @@ func (vi *ValueIndex) Drop(txn *badger.Txn, utxoID []byte) error {
 	return utils.DeleteValue(txn, key)
 }
 
-func (vi *ValueIndex) GetValueForOwner(txn *badger.Txn, owner *objs.Owner, minValue *uint256.Uint256, exclude [][]byte) ([][]byte, *uint256.Uint256, error) {
-	exclusionSet := make(map[[constants.HashLen]byte]bool)
-	var hsh [constants.HashLen]byte
-	for j := 0; j < len(exclude); j++ {
-		ID := utils.CopySlice(exclude[j])
-		copy(hsh[:], utils.CopySlice(ID))
-		exclusionSet[hsh] = true
-	}
-	result := [][]byte{}
-	valueCount := uint256.Zero()
-	opts := badger.DefaultIteratorOptions
-	prefix := vi.prefix()
+func (vi *ValueIndex) GetValueForOwner(txn *badger.Txn, owner *objs.Owner, minValue *uint256.Uint256, excludeFn func([]byte) (bool, error), maxCount int, lastKey []byte) ([][]byte, *uint256.Uint256, []byte, error) {
 	ownerBytes, err := owner.MarshalBinary()
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
+
+	prefix := vi.prefix()
 	prefix = append(prefix, ownerBytes...)
-	prefixLen := len(prefix)
+	opts := badger.DefaultIteratorOptions
 	opts.Prefix = prefix
 	iter := txn.NewIterator(opts)
 	defer iter.Close()
-	for iter.Seek(prefix); iter.ValidForPrefix(prefix); iter.Next() {
+
+	result := [][]byte{}
+	totalValue := uint256.Zero()
+	prefixLen := len(prefix)
+
+	if lastKey != nil {
+		iter.Seek(lastKey)
+		if !iter.ValidForPrefix(prefix) {
+			return result, totalValue, nil, nil
+		}
+		iter.Next()
+	} else {
+		iter.Seek(prefix)
+	}
+
+	for ; iter.ValidForPrefix(prefix); iter.Next() {
 		itm := iter.Item()
 		key := itm.KeyCopy(nil)
+
 		valueBytes := key[prefixLen : len(key)-constants.HashLen]
 		value := &uint256.Uint256{}
-		err := value.UnmarshalBinary(utils.CopySlice(valueBytes))
+		err := value.UnmarshalBinary(valueBytes)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
+
 		utxoID, err := itm.ValueCopy(nil)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
-		copy(hsh[:], utxoID)
-		if !exclusionSet[hsh] {
-			valueCount, err = valueCount.Clone().Add(valueCount.Clone(), value.Clone())
+
+		if excludeFn != nil {
+			shouldExclude, err := excludeFn(utxoID)
 			if err != nil {
-				return nil, nil, err
+				return nil, nil, nil, err
 			}
-			result = append(result, utxoID)
+			if shouldExclude {
+				continue
+			}
+		}
+
+		totalValue, err = totalValue.Add(totalValue, value)
+		if err != nil {
+			return nil, nil, nil, err
 		}
-		if valueCount.Gte(minValue) {
+
+		result = append(result, utxoID)
+
+		if totalValue.Gte(minValue) {
 			break
 		}
+		if len(result) >= maxCount {
+			return result, totalValue, key, nil
+		}
 	}
-	return result, valueCount.Clone(), nil
+	return result, totalValue, nil, nil
```

Comment:
Why not return token? I see the exit logic implies we left loop due to gt min value but providing token does not hurt in case client wants to apply different logic for utxo selection.